tests/petsc/iterate_parallel_01.cc

```diff
@@ -0,0 +1,114 @@
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// check re-initializing a preconditioner (parallel version)
+
+#include "../tests.h"
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/petsc_parallel_sparse_matrix.h>
+#include <deal.II/lac/petsc_precondition.h>
+#include <deal.II/base/index_set.h>
+#include <iostream>
+#include <vector>
+
+template <class PRE>
+void test ()
+{
+  unsigned int myid = Utilities::MPI::this_mpi_process (MPI_COMM_WORLD);
+  unsigned int numproc = Utilities::MPI::n_mpi_processes (MPI_COMM_WORLD);
+
+  deallog << numproc << std::endl;
+
+  // each processor owns 2 indices and all
+  // are ghosting Element 1 (the second)
+
+  IndexSet local_active(numproc*2);
+  local_active.add_range(myid*2,myid*2+2);
+  IndexSet local_relevant= local_active;
+  local_relevant.add_range(0,1);
+
+  DynamicSparsityPattern csp (local_relevant);
+
+  for (unsigned int i=0; i<2*numproc; ++i)
+    if (local_relevant.is_element(i))
+      csp.add(i,i);
+
+  csp.add(0,1);
+
+
+  PETScWrappers::MPI::SparseMatrix mat;
+  mat.reinit (local_active, local_active, csp, MPI_COMM_WORLD);
+
+  Assert(mat.n()==numproc*2, ExcInternalError());
+  Assert(mat.m()==numproc*2, ExcInternalError());
+
+  // set local values
+  mat.set(myid*2,myid*2, 1.0+myid*2.0);
+  mat.set(myid*2+1,myid*2+1, 2.0+myid*2.0);
+
+  mat.compress(VectorOperation::insert);
+
+  {
+    PETScWrappers::MPI::Vector src, dst;
+    src.reinit(local_active, MPI_COMM_WORLD);
+    dst.reinit(local_active, MPI_COMM_WORLD);
+    src(myid*2) = 1.0;
+    src.compress(VectorOperation::insert);
+
+    PRE pre;
+    pre.initialize(mat);
+    pre.vmult(dst, src);
+    dst.print(deallog.get_file_stream());
+
+    mat.add(0,0,1.0);
+    mat.compress(VectorOperation::add);
+
+    pre.initialize(mat);
+    pre.vmult(dst, src);
+    dst.print(deallog.get_file_stream());
+  }
+
+  //////////////////////////////////////////////
+  /////This is a test for the local matrix iterator
+  //////////////////////////////////////////////
+  unsigned int start_row = mat.local_range().first;
+  unsigned int end_row = mat.local_range().second;
+  for (auto r = start_row; r < end_row; ++r)
+    {
+      for (auto itr = mat.begin(r); itr != mat.end(r); ++itr)
+        {
+          //do nothing, just iterate
+        }
+    }
+  /////////////////////////////////////////////
+
+  if (myid==0)
+    deallog << "OK" << std::endl;
+}
+
+
+int main (int argc, char **argv)
+{
+  Utilities::MPI::MPI_InitFinalize mpi_initialization (argc, argv, 1);
+  MPILogInitAll log;
+
+  test<PETScWrappers::PreconditionJacobi> ();
+  test<PETScWrappers::PreconditionBlockJacobi> ();
+  test<PETScWrappers::PreconditionEisenstat> ();
+  test<PETScWrappers::PreconditionBoomerAMG> ();
+  test<PETScWrappers::PreconditionParaSails> ();
+  test<PETScWrappers::PreconditionNone> ();
```

Comment:
Same issue: we don't need to run the test over multiple preconditioners (their state does not impact the iterator).